src/query/api/v1/httpd/handler.go

```diff
@@ -318,19 +308,91 @@ func (h *Handler) RegisterRoutes() error {
 					Tagged(remoteSource).
 					Tagged(experimentalAPIGroup),
 			)
-			h.router.HandleFunc(annotated.WriteURL,
-				wrapped(experimentalAnnotatedWriteHandler).ServeHTTP,
-			).Methods(annotated.WriteHTTPMethod)
+			addWrappedRoute(annotated.WriteURL, experimentalAnnotatedWriteHandler,
+				annotated.WriteHTTPMethod)
 		}
 	}
 
 	h.registerHealthEndpoints()
 	h.registerProfileEndpoints()
 	h.registerRoutesEndpoint()
 
+	// Register custom endpoints.
+	for _, custom := range h.customHandlers {
+		routeName := routeName(custom.Route(), custom.Methods())
+		route := h.router.Get(routeName)
+		var prevHandler http.Handler
+		if route != nil {
+			prevHandler = route.GetHandler()
+		}
+		customHandler, err := custom.Handler(nativeSourceOpts, prevHandler)
+		if err != nil {
+			return fmt.Errorf("failed to register custom handler with path %s: %w",
+				routeName, err)
+		}
+
+		if route == nil {
+			addWrappedRoute(custom.Route(), customHandler, custom.Methods()...)
+		} else {
+			route.Handler(wrapped(customHandler))
+		}
+	}
+
 	return nil
 }
 
+func addRoute(router *mux.Router, path string, handler http.Handler, methods []string) {
+	addRouteHandlerFn(router, path, handler.ServeHTTP, methods...)
+}
+
+func addRouteHandlerFn(
+	router *mux.Router,
+	path string,
+	handlerFn func(http.ResponseWriter, *http.Request),
+	methods ...string,
+) {
+	routeName := routeName(path, methods)
+	if previousRoute := router.Get(routeName); previousRoute != nil {
+		// route already exists.
+		return
+	}
+
+	route := router.
+		HandleFunc(path, handlerFn).
+		Name(routeName)
+
+	if len(methods) > 0 {
+		route.Methods(methods...)
+	}
+}
+
+func routeName(p string, methods []string) string {
```

Comment:
At least to me, this seems to put a strict requirements on how the handlers are registered. In order to successfully override a handler, one needs to precisely match implementation details:
- path has to be exactly the same (e.g. `/api/v1/label/{name}/values` and `/api/v1/label/{label}/values` are semantically the same, but they won't trigger override code branch);
- the same methods have to be registered in the same way (e.g. one might register the same handler twice, `GET /resource` and `POST /resource`, and override wouldn't work if the other side would do the same in a single call, i.e. `(GET|POST) /resource`).

These things have to match during initial implementation and the maintained during code changes.


Also, what should be the behavior if base handler registers `(GET|POST) /resource`, but custom handler is `GET /resource`? What if it's vice-versa? In both of these cases there's an overlap.

Reply:
Yeah, I'm aware of this. The current implementation is not finished yet. Actually, I was thinking to register a separate route for each method, so if one handler wants to add a new route with 2 methods (e.g. GET and POST), 2 named routes would be added (one for GET and another one for POST). Then custom handler could override even single method for a previously registered route with multiple methods. Does this makes sense?

Reply:
👍 that should make the behavior much more intuitive